plugins/audio-s5gen2/fu-audio-s5gen2-device.c

```diff
@@ -0,0 +1,724 @@
+	FuDevice *proxy = fu_device_get_proxy(FU_DEVICE(self));
+
+	if (klass->msg_cmd == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "not implemented (%s)",
+			    __FUNCTION__);
+		return FALSE;
+	}
+
+	return klass->msg_cmd(proxy, data, data_len, error);
+}
+
+static gboolean
+fu_qc_s5gen2_device_cmd_req_disconnect(FuQcS5gen2Device *self, GError **error)
+{
+	g_autoptr(GByteArray) req = fu_struct_qc_disconnect_req_new();
+
+	return fu_qc_s5gen2_device_msg_cmd(self, req->data, req->len, error);
+}
+
+static gboolean
+fu_qc_s5gen2_device_cmd_req_connect(FuQcS5gen2Device *self, GError **error)
+{
+	guint8 data_in[FU_STRUCT_QC_UPDATE_STATUS_SIZE] = {0x0};
+	FuQcStatus update_status;
+	g_autoptr(GByteArray) req = fu_struct_qc_connect_req_new();
+	g_autoptr(GByteArray) st = NULL;
+
+	if (!fu_qc_s5gen2_device_msg_cmd(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_device_msg_in(self, data_in, sizeof(data_in), error))
+		return FALSE;
+
+	st = fu_struct_qc_update_status_parse(data_in, sizeof(data_in), 0, error);
+	if (st == NULL)
+		return FALSE;
+
+	update_status = fu_struct_qc_update_status_get_status(st);
+	switch (update_status) {
+	case FU_QC_STATUS_SUCCESS:
+		break;
+	case FU_QC_STATUS_ALREADY_CONNECTED_WARNING:
+		g_info("device is already connected");
+		/* FIXME: continue the previous update for wireless
+		 * atm fail for USB */
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "device is already connected");
+		return FALSE;
+	default:
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid update status (%s)",
+			    fu_qc_status_to_string(update_status));
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_device_cmd_abort(FuQcS5gen2Device *self, GError **error)
+{
+	guint8 data[FU_STRUCT_QC_ABORT_SIZE] = {0};
+	g_autoptr(GByteArray) req = fu_struct_qc_abort_req_new();
+	g_autoptr(GByteArray) reply = NULL;
+
+	if (!fu_qc_s5gen2_device_msg_out(self, req->data, req->len, error))
+		return FALSE;
+	if (!fu_qc_s5gen2_device_msg_in(self, data, sizeof(data), error))
+		return FALSE;
+
+	reply = fu_struct_qc_abort_parse(data, sizeof(data), 0, error);
+	if (reply == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_device_cmd_sync(FuQcS5gen2Device *self, GError **error)
+{
+	FuQcS5gen2DevicePrivate *priv = GET_PRIVATE(self);
+	guint8 data[FU_STRUCT_QC_SYNC_SIZE] = {0};
+	FuQcResumePoint rp;
+	g_autoptr(GByteArray) req = fu_struct_qc_sync_req_new();
+	g_autoptr(GByteArray) reply = NULL;
+
+	fu_struct_qc_sync_req_set_file_id(req, priv->file_id);
+	if (!fu_qc_s5gen2_device_msg_out(self, req->data, req->len, error))
+		return FALSE;
+	if (!fu_qc_s5gen2_device_msg_in(self, data, sizeof(data), error))
+		return FALSE;
+
+	/* FIXME: correct error handling -- move to msg_in()? */
+	if (data[0] == 0x11) {
+		g_set_error(error,
+			    G_IO_ERROR,
```

Comment:
switch this and all other `G_IO_ERROR` to the new matching `FWUPD` quark errors (see your precommit failure for details).

Reply:
Yes, I 've seen the error already.